src/lib.rs

```
@@ -0,0 +1,93 @@
+use lazy_static::lazy_static;
+use std::collections::HashSet;
+
+use guest::prelude::*;
+use kubewarden_policy_sdk::wapc_guest as guest;
+
+use k8s_openapi::api::core::v1 as apicore;
+
+extern crate kubewarden_policy_sdk as kubewarden;
+use kubewarden::{logging, protocol_version_guest, request::ValidationRequest, validate_settings};
+
+mod settings;
+use settings::Settings;
+
+use slog::{error, o, warn, Logger};
+
+lazy_static! {
+    static ref LOG_DRAIN: Logger = Logger::root(
+        logging::KubewardenDrain::new(),
+        o!("policy" => "priority-class-policy")
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn wapc_init() {
+    register_function("validate", validate);
+    register_function("validate_settings", validate_settings::<Settings>);
+    register_function("protocol_version", protocol_version_guest);
+}
+
+fn validate(payload: &[u8]) -> CallResult {
+    let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
+    match validation_request.extract_pod_spec_from_object() {
+        Ok(Some(pod_spec)) => {
+            match validate_pod_priority_class(
+                pod_spec,
+                validation_request.settings.allowed_priority_classes,
+            ) {
+                Ok(_) => kubewarden::accept_request(),
+                Err(err) => kubewarden::reject_request(Some(err.to_owned()), None, None, None),
+            }
+        }
+        Ok(None) => {
+            warn!(LOG_DRAIN, "no PodSpec found");
+            kubewarden::accept_request()
+        }
+        Err(_) => {
+            error!(LOG_DRAIN, "cannot extract PodSpec");
```

please, do not swallow the error, add it to the log

---

I've added the error in the warnings fields:

```
  Successfully pulled policy from file:///home/jvanz/SUSE/priority-class-policy/annotated-policy.wasm
2025-04-30T11:58:50.379510Z  WARN validate{self=PolicyEvaluator { runtime: "wapc" } settings={"allowed_priority_classes": Array [String("noncritical")]}}:policy_log{self=EvaluationContext { policy_id: "priority-class-policy", callback_channel: Some(...), allowed_kubernetes_resources: {} }}: policy_log: cannot extract PodSpec data={"column":13,"file":"src/lib.rs","line":48,"policy":"priority-class-policy"}
{"uid":"705ab4f5-6393-11e8-b7cc-42010a800002","allowed":true,"auditAnnotations":null,"warnings":["Priority class policy failed to extract PodSpec from the request: Object should be one of these kinds: Deployment, ReplicaSet, StatefulSet, DaemonSet, ReplicationController, Job, CronJob, Pod"]}

```